File: cores_dec.py
import heapq
import igraph
import copy

__author__ = "antoine.tixier-1@colorado.edu"

def core_dec(g, weighted = True):
    # unweighted or weighted k-core decomposition of a graph
  
    # work on clone of g to preserve g 
    gg = copy.deepcopy(g)    
    
    # initialize dictionary that will contain the core numbers
    cores_g = dict(zip(gg.vs["name"],[0]*len(gg.vs["name"])))
    
    if weighted == True:
        # k-core decomposition for weighted graphs (generalized k-cores)
        # based on Batagelj and Zaversnik's (2002) algorithm #4
    
        # initialize min heap of degrees
        heap_g = zip(gg.vs["weight"],gg.vs["name"])
        heapq.heapify(heap_g)
        
        while len(heap_g)>0:
            
            top = heap_g[0][1]
            # find vertice index of heap top element
            index_top = gg.vs["name"].index(top)
            # save names of its neighbors
            neighbors_top = gg.vs[gg.neighbors(top)]["name"]
            # set core number of heap top element as its weighted degree
            cores_g[top] = gg.vs["weight"][index_top]
            # delete top vertice (weighted degrees are automatically updated)
            gg.delete_vertices(index_top)
            
            if len(neighbors_top)>0:
            # iterate over neighbors of top element
                for i, name_n in enumerate(neighbors_top):
                    index_n = gg.vs["name"].index(name_n)
                    max_n = max(cores_g[top],gg.strength(weights=gg.es["weight"])[index_n])
                    gg.vs[index_n]["weight"] = max_n
                    # update heap
                    heap_g = zip(gg.vs["weight"],gg.vs["name"])
                    heapq.heapify(heap_g)
            else:
                # update heap
                heap_g = zip(gg.vs["weight"],gg.vs["name"])
                heapq.heapify(heap_g)
                
    else:
        # k-core decomposition for unweighted graphs
        # based on Batagelj and Zaversnik's (2002) algorithm #1
    
        # compute vertice degrees and order vertices in increasing order of their degrees
        degrees_g = sorted(zip(gg.vs["name"],gg.strength()), key=operator.itemgetter(1))
        
        for i, degrees_g_temp in enumerate(degrees_g):
            vertice = degrees_g_temp[0]
            cores_g[vertice] =  degrees_g_temp[1]
            neighbors_vertice = gg.vs[gg.neighbors(vertice)]["name"]
            for j, neighbor in enumerate(neighbors_vertice):
                index_neighbor = [temp_tuple[0] for temp_tuple in degrees_g].index(neighbor)
                if degrees_g[index_neighbor][1] > degrees_g[i][1]:
                    degrees_g[index_neighbor] = (neighbor, degrees_g[index_neighbor][1] - 1)
                    # reorder vertices in increasing order of their degrees
                    degrees_g = sorted(degrees_g, key=operator.itemgetter(1))
    
    # sort vertices by decreasing core number
    sorted_cores_g = sorted(cores_g.items(), key=operator.itemgetter(1), reverse=True)
    
    return(sorted_cores_g)

File: cores_dec.R
cores_dec = function(g, weighted){
  # This function performs unweighted or weighted k-core decomposition of a graph of terms.
  # The graph should be undirected and (optionally) edge-weighted
  # The function returns an ordered list of 2-tuples where each term is associated with its core number
  
  # With respect to keyword extraction:
  # - select unweighted k-core when recall is more important than precision (larger main core)
  # - select weighted k-core when precision is more important than recall (smaller but stronger main core)
  # - note that overall (in terms of F1 score), weighted k-core performs better
  
  # initialize list that will contain the core numbers
  cores_g = 1:length(V(g)$name)*0
  names(cores_g) = V(g)$name
  
  if (weighted == TRUE){
    # k-core decomposition for weighted graphs (generalized k-cores)
    # based on Batagelj and Zaversnik's (2002) algorithm #4
    
    # create and heapify list of initial weighted degrees (nodes are alphabetically ordered for consistency with Python)
    V(g)$weight = strength(g, mode="all")
    heap_g = V(g)$weight
    names(heap_g) = V(g)$name
    heap_g=heap_g[order(names(heap_g))]
    heap_g=heapify(heap_g, min=TRUE)$heap
    
    while (length(heap_g)>0) {
      
      top = heap_g[1]
      name_top = names(top)
      # save hop-1 neighborhood of heap top element (vertice of minimum weighted degree)
      neighbors_top = setdiff(names(neighborhood(g, order=1, nodes=name_top)[[1]]), name_top)
      # order neighbors by alphabetical order (for consistency with Python)
      neighbors_top = sort(neighbors_top)
      
      # set core number of heap top element as its value (weighted degree)
      cores_g[name_top] = V(g)$weight[which(V(g)$name==name_top)]
      # delete top vertice from graph
      g = delete_vertices(g, name_top)
      
      gstr = strength(g, mode="all")
      
      if (length(neighbors_top)>0){
        # iterate over neighbors of top element
        for (i in 1:length(neighbors_top)){
          name_n = neighbors_top[i]
          max_n = max(cores_g[name_top], gstr[name_n])
          V(g)$weight[which(V(g)$name==name_n)] = max_n
         
          # update heap
          heap_g = V(g)$weight
          names(heap_g) = V(g)$name
          heap_g=heap_g[order(names(heap_g))]
          heap_g=heapify(heap_g, min=TRUE)$heap

        }
      } else {
        # update heap
        heap_g = V(g)$weight
        names(heap_g) = V(g)$name
        heap_g=heap_g[order(names(heap_g))]
        heap_g=heapify(heap_g, min=TRUE)$heap
      }
      
    }
    
  } else {
    # k-core decomposition for unweighted graphs
    # based on Batagelj and Zaversnik's (2002) algorithm #1
    cores_g = graph.coreness(g)
  }
  
  # sort vertices by decreasing core number
  cores_g = sort(cores_g, decreasing = TRUE)
  output = list(cores = cores_g)
}
